### PR TITLE
C#: Add support for custom assert methods (DoesNotReturnIfAttribute)

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

**Comment by tamasvajk** [comment]
todo: add change notes.

File: csharp/ql/src/semmle/code/csharp/commons/Assertions.qll [hunk 1]
```diff
@@ -9,11 +9,13 @@ private import ControlFlow::BasicBlocks
 
 /** An assertion method. */
 abstract class AssertMethod extends Method {
-  /** Gets the index of the parameter being asserted. */
-  abstract int getAssertionIndex();
+  /** Gets the index of a parameter being asserted. */
+  abstract int getAnAssertionIndex();
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Assertions.qll:13]: We should add deprecated `getAssertionIndex()`, `getAssertedParameter()`, and `getExpr()` predicates for backwards compatibility.

File: csharp/ql/src/semmle/code/csharp/commons/Assertions.qll [hunk 2]
```diff
@@ -195,11 +197,60 @@ class SystemDiagnosticsContractAssertTrueMethod extends AssertTrueMethod {
   }
 }
 
+private predicate isDoesNotReturnIfAttributeParameter(Parameter p, boolean value) {
+  exists(Attribute a | a = p.getAnAttribute() |
+    a.getType() = any(SystemDiagnosticsCodeAnalysisDoesNotReturnIfAttributeClass c) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Assertions.qll:202]:
```suggestion
    a.getType() instanceof SystemDiagnosticsCodeAnalysisDoesNotReturnIfAttributeClass and
```

File: csharp/ql/src/semmle/code/csharp/commons/Assertions.qll [hunk 3]
```diff
@@ -195,11 +197,60 @@ class SystemDiagnosticsContractAssertTrueMethod extends AssertTrueMethod {
...
+class SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod extends AssertTrueMethod {
+  SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod() {
+    this = any(Method m | isDoesNotReturnIfAttributeParameter(m.getAParameter(), false))
+  }
+
+  override int getAnAssertionIndex() {
+    exists(Parameter p |
+      this.getParameter(result) = p and isDoesNotReturnIfAttributeParameter(p, false)
+    )
+  }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Assertions.qll:220]:
```suggestion
  private int i;

  SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod() {
    isDoesNotReturnIfAttributeParameter(this.getParameter(i), false)
  }

  override int getAnAssertionIndex() { result = i }
```

File: csharp/ql/src/semmle/code/csharp/commons/Assertions.qll [hunk 4]
```diff
@@ -195,11 +197,60 @@ class SystemDiagnosticsContractAssertTrueMethod extends AssertTrueMethod {
...
+
+  override Class getExceptionClass() {
+    // The user defines the thrown exception.
+    any()
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Assertions.qll:224]: This creates a large Cartesian product, so better to return the `System.Exception` base class.

File: csharp/ql/src/semmle/code/csharp/commons/Assertions.qll [hunk 5]
```diff
@@ -195,11 +197,60 @@ class SystemDiagnosticsContractAssertTrueMethod extends AssertTrueMethod {
...
+  }
+}
+
+/**
+ * A method with a parameter that is annotated with
+ * `System.Diagnostics.CodeAnalysis.DoesNotReturnIfAttribute(true)`.
+ */
+class SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertFalseMethod extends AssertFalseMethod {
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Assertions.qll:232]: Same comments as for `SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod`.

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Splitting.qll [hunk 6]
```diff
@@ -445,7 +445,7 @@ module AssertionSplitting {
 
     override predicate hasEntry(ControlFlowElement pred, ControlFlowElement succ, Completion c) {
       exists(AssertMethod m |
-        pred = last(a.getExpr(), c) and
+        pred = last(a.getAnExpr(), c) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/controlflow/internal/Splitting.qll:448]: The CFG splitting logic will not work for assertions with multiple assertion arguments; I will address in a follow-up PR.

**Review by hvitved — CHANGES_REQUESTED** [review]

**Comment by tamasvajk** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/568/

**Comment by tamasvajk** [comment]
> https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/568/

There are some nice fixes in the results, for example, we're not reporting on the following that it can be null: https://github.com/dotnet/runtime/blob/45833cb6e43b481cdb1dabe2ce29fca48ee7bf6a/src/libraries/System.Linq.Expressions/src/System/Dynamic/DynamicObject.cs#L465

**Comment by hvitved** [comment]
> There are some nice fixes in the results, for example, we're not reporting on the following that it can be null: https://github.com/dotnet/runtime/blob/45833cb6e43b481cdb1dabe2ce29fca48ee7bf6a/src/libraries/System.Linq.Expressions/src/System/Dynamic/DynamicObject.cs#L465

That's excellent!

**Review by hvitved — APPROVED** [review]

**Pull request merged**